doc: suit: user guide fixes for 2.7.0 (#15804)
nrfconnect/sdk-nrf: doc/nrf/device_guides/working_with_nrf/nrf54h/ug_nrf54h20_suit_customize_dfu.rst

Build system configuration
**************************

By default the build system generates SUIT envelopes using predefined manifest templates provided by Nordic Semiconductor.

> These templates, suitable for standard development needs, can be found in :file:`modules/lib/suit-generator/ncs`

Reviewer: ```suggestion
These templates can be found in :file:`modules/lib/suit-generator/ncs`, and are suitable for standard development needs.
```

> These templates, suitable for standard development needs, can be found in :file:`modules/lib/suit-generator/ncs`

Reviewer: ```suggestion
These templates can be found in :file:`modules/lib/suit-generator/ncs`, and are suitable for standard development needs.
```

.

Three manifests are used in the most common case:

> * Root manifest - :file:`root_hierarchical_envelope.yaml.jinja2`

Reviewer 2: I cannot find this file inside suit-generator repo

* Application domain manifest - :file:`app_envelope.yaml.jinja2`
* Radio domain manifest - :file:`rad_envelope.yaml.jinja2`

.. note::
   The radio domain manifest template is available only for the Bluetooth® Low Energy version of the :ref:`nrf54h_suit_sample`, not the UART version.

> The whole process of building a DFU envelope (which contains the manifests) can be summarized with the following diagram:

Reviewer: ```suggestion
The process of building a SUIT envelope (which contains the manifests) can be summarized with the following diagram:
```

.. figure:: images/nrf54h20_suit_generator_workflow.png
   :alt: Modifying manifest templates workflow

   Modifying manifest templates workflow

Manifest templates (``.jinja2``) files are rendered to YAML files.
These YAML files are used as a representation of the output binary SUIT envelopes.
The provided manifest templates (``.jinja2``) files contain variables (represented as components), such as memory ranges, names, and paths to binaries.
The component values are filled out automatically by the build system during the manifest rendering.

Variables in the provided templates, like memory ranges and paths to binaries, are filled out by the build system.

.. _ug_suit_use_own_manifest:

How to use your own manifest
----------------------------

Changing the source of manifest templates allows you to create manifest files on your own, without relying on the templates provided by Nordic Semiconductor.

The source of the manifest templates can be configured by setting the following Kconfig options:

* :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE_ROOT_TEMPLATE`

* :kconfig:option:`CONFIG_SUIT_ENVELOPE_TEMPLATE` - for each of the images (application and radio core images)

> Let us analyze the following case:

Reviewer: ```suggestion
One example is demonstrated with the following case:
```

.. tabs::

    .. group-tab:: Windows

>         The provided user-defined manifest templates are stored in ``C:\my_default_templates``.

Reviewer 2: I'm not sure if that works in command-line parameters, but from Kconfig you can also use CMake variables to manage those paths, i.e. `$(ZEPHYR_NRF_MODULE_DIR)`.
Feel free to test&update it if that simplifies the docs (i.e. results in a common code for all OSes)

    .. group-tab:: Linux

        The provided user-defined manifest templates are stored in ``/home/my_user/my_default_templates``.

> The following files are used to create DFU envelope:

Reviewer: ```suggestion
The following files are used to create the SUIT envelope:
```

      .. code-block:: console

         west build --sysbuild -d C:/ncs-lcs/work_dir/build/ -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_SUIT_ENVELOPE_EDITABLE_TEMPLATES_LOCATION="C:/my_templates"

   .. group-tab:: Linux

      Enter the following command to change the location where the editable manifest templates are stored:

      .. code-block:: console

         west build --sysbuild -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_SUIT_ENVELOPE_EDITABLE_TEMPLATES_LOCATION=\"/home/my_user/my_templates\"

The source of the manifest templates can be configured by setting the following Kconfig options:

* :kconfig:option:`SB_CONFIG_SUIT_ENVELOPE_ROOT_TEMPLATE`

> * :kconfig:option:`CONFIG_SUIT_ENVELOPE_TEMPLATE` - for each of the images (application/radio core)

Reviewer: ```suggestion
* :kconfig:option:`CONFIG_SUIT_ENVELOPE_TEMPLATE` - for each of the images (application and radio core images)
```

Author: Fixed

        .. code-block:: console

>             west build --sysbuild -d C:/ncs-lcs/work_dir/build/ -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_SUIT_ENVELOPE_ROOT_TEMPLATE="c:/my_default_templates/root.yaml.jinja2" -DCONFIG_SUIT_ENVELOPE_TEMPLATE="c:/my_default_templates/app.yaml.jinja2" -Dhci_ipc_CONFIG_SUIT_ENVELOPE_SUIT_ENVELOPE_TEMPLATE="c:/my_default_templates/radio.yaml.jinja2"

Reviewer 3: ```suggestion
            west build --sysbuild -d C:/ncs-lcs/work_dir/build/ -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_SUIT_ENVELOPE_ROOT_TEMPLATE="c:/my_default_templates/root.yaml.jinja2" -DCONFIG_SUIT_ENVELOPE_TEMPLATE="c:/my_default_templates/app.yaml.jinja2" -Dhci_ipc_CONFIG_SUIT_ENVELOPE_TEMPLATE="c:/my_default_templates/radio.yaml.jinja2"
```

Author: Fixed

    .. group-tab:: Linux

        Run the following command:

        .. code-block:: console

>             west build --sysbuild -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_SUIT_ENVELOPE_ROOT_TEMPLATE=\"/home/my_user/my_default_templates/root.yaml.jinja2\" -DCONFIG_SUIT_ENVELOPE_TEMPLATE=\"/home/my_user/my_default_templates/app.yaml.jinja2\" -Dhci_ipc_CONFIG_SUIT_ENVELOPE_SUIT_ENVELOPE_TEMPLATE=\"/home/my_user/my_default_templates/radio.yaml.jinja2\"

Reviewer 3: ```suggestion
            west build --sysbuild -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_SUIT_ENVELOPE_ROOT_TEMPLATE=\"/home/my_user/my_default_templates/root.yaml.jinja2\" -DCONFIG_SUIT_ENVELOPE_TEMPLATE=\"/home/my_user/my_default_templates/app.yaml.jinja2\" -Dhci_ipc_CONFIG_SUIT_ENVELOPE_SUIT_ENVELOPE_TEMPLATE=\"/home/my_user/my_default_templates/radio.yaml.jinja2\"
```
```suggestion
            west build --sysbuild -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_SUIT_ENVELOPE_ROOT_TEMPLATE=\"/home/my_user/my_default_templates/root.yaml.jinja2\" -DCONFIG_SUIT_ENVELOPE_TEMPLATE=\"/home/my_user/my_default_templates/app.yaml.jinja2\" -Dhci_ipc_CONFIG_SUIT_ENVELOPE_TEMPLATE=\"/home/my_user/my_default_templates/radio.yaml.jinja2\"
```

Author: Fixed

        .. code-block:: console

            west build -d C:/ncs-lcs/work_dir/build/ -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_SUIT_ENVELOPE_ROOT_TEMPLATE="c:/my_default_templates/root.yaml.jinja2" -DCONFIG_SUIT_ENVELOPE_TEMPLATE="c:/my_default_templates/app.yaml.jinja2" -Dhci_ipc_CONFIG_SUIT_ENVELOPE_TEMPLATE="c:/my_default_templates/radio.yaml.jinja2"

    .. group-tab:: Linux

        Run the following command:

        .. code-block:: console

            west build -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_SUIT_ENVELOPE_ROOT_TEMPLATE=\"/home/my_user/my_default_templates/root.yaml.jinja2\" -DCONFIG_SUIT_ENVELOPE_TEMPLATE=\"/home/my_user/my_default_templates/app.yaml.jinja2\" -Dhci_ipc_CONFIG_SUIT_ENVELOPE_TEMPLATE=\"/home/my_user/my_default_templates/radio.yaml.jinja2\"

For more information about the Kconfig options used in this example, see the `SUIT Kconfig options <https://res.developer.nordicsemi.com/ncs/doc/latest/kconfig/index.html#!suit_envelope>`__.

Customize Vendor and Class UUIDs
----------------------------

Customizing UUIDs used for class and vendor IDs enhances security and is recommended for specific use cases.

> Values of ``class-identifier`` and ``vendor-identifier`` in the manifest are created based on the ``CONFIG_SUIT_MPI_<MANIFEST_ROLE>_VENDOR_NAME`` and ``CONFIG_SUIT_MPI_<MANIFEST_ROLE>_CLASS_NAME`` options.

Reviewer: ```suggestion
Values for ``class-identifier`` and ``vendor-identifier`` in the manifest are created based on the ``CONFIG_SUIT_MPI_<MANIFEST_ROLE>_VENDOR_NAME`` and ``CONFIG_SUIT_MPI_<MANIFEST_ROLE>_CLASS_NAME`` Kconfig options.
```

        .. code-block:: console

>             west build --sysbuild -d C:/ncs-lcs/work_dir/build/ -b nrf54h20dk/nrf54h20/cpuapp -p -- -DSB_CONFIG_SUIT_ENVELOPE_EDITABLE_TEMPLATES_LOCATION="c:/my_templates" -DSB_CONFIG_SUIT_ENVELOPE_ROOT_TEMPLATE="c:/my_default_templates/root.yaml.jinja2" -DCONFIG_SUIT_ENVELOPE_TEMPLATE="c:/my_default_templates/app.yaml.jinja2" -Dhci_ipc_CONFIG_SUIT_ENVELOPE_SUIT_ENVELOPE_TEMPLATE="c:/my_default_templates/radio.yaml.jinja2"

Author: Here as well as for the other examples - how do we escape the paths for `CONFIG_SUIT_ENVELOPE_TEMPLATE` (and other paths) on Windows? This needs to be tested

Additionally, the Python dictionary holds a variable called ``version`` that holds the application version.
With the Python dictionary you are able to, for example:

* Extract the CPU ID by using ``application['dt'].label2node['cpu'].unit_addr``

* Obtain the partition address with ``application['dt'].chosen_nodes['zephyr,code-partition']``

* Obtain the size of partition with ``application['dt'].chosen_nodes['zephyr,code-partition'].regs[0].size``

* Get the pair of URI name and the binary path by using ``'#{{ application['name'] }}': {{ application['binary'] }}``

> * Get the application version with ``suit-manifest-sequence-number: {{ version }}``

Reviewer 2: SHouldn't that be `sysbuild['config']['SB_CONFIG_SUIT_ENVELOPE_SEQUENCE_NUM']`?

Set component definition and memory ranges
------------------------------------------

> In the :file:`app_envelope_yaml.jinja2` (found `here <https://github.com/nrfconnect/suit-generator/blob/v2.7.0/ncs/app_envelope.yaml.jinja2>`__), the component definition and memory ranges are filled out by using the ``edtlib`` (devicetree values) object like so:

Reviewer 3: Is it okay that we are using a link here? The page doesn't exist yet.
Maybe it should point to this instead?
```
`:file:`modules/lib/suit-generator/ncs/app_envelope.yaml.jinja2`
```

Author: Changed to "file"

Set component definition and memory ranges
------------------------------------------

In :file:`modules/lib/suit-generator/ncs/app_envelope.yaml.jinja2`
, the component definition and memory ranges are filled out by using the ``edtlib`` (devicetree values) object like so:

.. code-block::

    suit-components:
    - - MEM
    - ``{{ application['dt'].label2node['cpu'].unit_addr }}``
    - ``{{ get_absolute_address(application['dt'].chosen_nodes['zephyr,code-partition']) }}``
    - ``{{ application['dt'].chosen_nodes['zephyr,code-partition'].regs[0].size }}``

.. note::
   See the :ref:`ug_suit_dfu_component_def` page for a full list and table of the available customizable components.

Set integrated payload
----------------------

> In :file:`modules/lib/suit-generator/ncs/app_envelope.yaml.jinja2`
> , the integrated payload definition is done using the target name and binary location:

Reviewer: ```suggestion
In :file:`modules/lib/suit-generator/ncs/app_envelope.yaml.jinja2`, the integrated payload definition is done using the target name and binary location:
```

           suit-digest-algorithm-id: cose-alg-sha-256
      suit-manifest:
         suit-manifest-version: 1

>          sysbuild['config']['SB_CONFIG_SUIT_ENVELOPE_SEQUENCE_NUM'] %}

Reviewer 2: Something is off here.

Some of these values are stored in the Python dictionaries that are named after the target name.
(Therefore, Python is used within the ``.jinja2`` files to fill in the necessary values in the manifest(s).)
For example, for the :ref:`nrf54h_suit_sample` there will be two variables available: ``application`` and ``radio``.

> The target names (the names of these variables) can be changed using ``CONFIG_SUIT_ENVELOPE_TARGET`` for a given image.

Reviewer: ```suggestion
The target names (the names of these variables) can be changed using the :kconfig:option:`CONFIG_SUIT_ENVELOPE_TARGET` Kconfig option for a given image.
```

Author: Fixed

Set component definition and memory ranges
------------------------------------------

> In the :file:`app_envelope_yaml.jinja2` (found `here <https://github.com/nrfconnect/suit-generator/blob/v2.7.0/ncs/app_envelope.yaml.jinja2>`__), the component definition and memory ranges are filled out by using the ``edtlib`` object like so:

Author: What is `edtlib` ? It isn't explained anywher, I do not understand what it is at nor will the user. Shouldn't we delete this information?

Reviewer: Earlier in the same doc, we mention this: https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/device_guides/working_with_nrf/nrf54h/ug_nrf54h20_suit_customize_dfu.html#variables_and_methods_available_in_the_manifest_templates
And provide a link to [edtlib](https://python-devicetree.readthedocs.io/en/latest/edtlib.html).

I think this is a valid point you've raised - if it is still unclear what ``edtib`` is, then we can add more info to the section where I first linked to, perhaps?

Author: Ok, I think it's ok if i just add the comment in the bracket that this is about devicetree values, then it is ok for me.

.. code-block::

    suit-components:
    - - MEM
    - ``{{ application['dt'].label2node['cpu'].unit_addr }}``
    - ``{{ get_absolute_address(application['dt'].chosen_nodes['zephyr,code-partition']) }}``
    - ``{{ application['dt'].chosen_nodes['zephyr,code-partition'].regs[0].size }}``

.. note::
   See the :ref:`ug_suit_dfu_component_def` page for a full list and table of the available customizable components.

Set integrated payload
----------------------

> In the :file:`app_envelope_yaml.jinja2` (found `here <https://github.com/nrfconnect/suit-generator/blob/v2.7.0/ncs/app_envelope.yaml.jinja2>`__, the integrated payload definition is done using the target name and binary location:

Author: @robertstypa  A reminder as we discussed - we will need a tag v2.7.0 in the suit-generator
